tool/tsh/common/scopes.go

```diff
@@ -0,0 +1,174 @@
+				PageToken: cursor,
+				// note that we are using the AllCallerAssignments flag here rather than just looking for
+				// our assignments by username. This flag suppresses standard scope-pinning, which allows
+				// us to see assignments in parent/orthogonal scopes. Generally, scoped commands only show
+				// the subset of state subject to the currently pinned scope, but the purpose of 'tsh scopes ls'
+				// is specifically to discover potential 'tsh login --scope=...' targets, so we want to see
+				// everything regardless of current scope.
+				AllCallerAssignments: true,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			assignments = append(assignments, rsp.Assignments...)
+			cursor = rsp.NextPageToken
+			if cursor == "" {
+				break
+			}
+		}
+		return nil
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if c.verbose {
+		// in verbose mode, we produce a table showing each scope and the roles
+		// assigned at that scope so that users can get a sense of the specific
+		// privileges they have at each scope.
+		type assignmentTableRow struct {
+			scope string
+			roles []string
+		}
+
+		var rows []*assignmentTableRow
+		for _, assignment := range assignments {
+		SubAssignments:
+			for _, subAssignment := range assignment.GetSpec().GetAssignments() {
+				for _, row := range rows {
+					if scopes.SortCmp(row.scope, subAssignment.GetScope()) == 0 {
+						row.roles = append(row.roles, subAssignment.GetRole())
+						continue SubAssignments
+					}
+				}
+				rows = append(rows, &assignmentTableRow{
+					scope: subAssignment.GetScope(),
+					roles: []string{subAssignment.GetRole()},
+				})
+			}
+		}
+
+		// scope sort order differs subtly from lexographic order, so we need a custom sort func here
+		slices.SortFunc(rows, func(a, b *assignmentTableRow) int {
+			return scopes.SortCmp(a.scope, b.scope)
+		})
+
+		table := asciitable.MakeTable([]string{"Scope", "Roles"})
+		for _, row := range rows {
+			slices.Sort(row.roles)
+			table.AddRow([]string{
+				row.scope,
+				strings.Join(row.roles, ", "),
+			})
+		}
+
+		fmt.Fprint(cf.Stdout(), table.AsBuffer().String())
+	} else {
+		// when not in verbose mode, the default output is a simple newline-delimited
+		// list of scopes. this keeps things simple, and has the nice upside of producing
+		// an output compatible with visualization tools like `tree` when piped appropriately.
+		var assignedScopes []string
+		for _, assignment := range assignments {
+			for _, subAssignment := range assignment.GetSpec().GetAssignments() {
+				assignedScopes = append(assignedScopes, subAssignment.GetScope())
+			}
+		}
+
+		// apply canonical sorting and deduplication
+		slices.SortFunc(assignedScopes, scopes.SortCmp)
+		assignedScopes = slices.CompactFunc(assignedScopes, func(a, b string) bool {
+			return scopes.SortCmp(a, b) == 0
+		})
+
+		for _, scope := range assignedScopes {
+			fmt.Fprintln(cf.Stdout(), scope)
+		}
+	}
```

Comment:
```suggestion

		fmt.Fprint(cf.Stdout(), table.AsBuffer().String())
		return nil
	}
	// when not in verbose mode, the default output is a simple newline-delimited
	// list of scopes. this keeps things simple, and has the nice upside of producing
	// an output compatible with visualization tools like `tree` when piped appropriately.
	var assignedScopes []string
	for _, assignment := range assignments {
		for _, subAssignment := range assignment.GetSpec().GetAssignments() {
			assignedScopes = append(assignedScopes, subAssignment.GetScope())
		}
	}

	// apply canonical sorting and deduplication
	slices.SortFunc(assignedScopes, scopes.SortCmp)
	assignedScopes = slices.CompactFunc(assignedScopes, func(a, b string) bool {
		return scopes.SortCmp(a, b) == 0
	})

	for _, scope := range assignedScopes {
		fmt.Fprintln(cf.Stdout(), scope)
	}
```
nit: can drop the else and indention in exchange for an extra `return nil`